File: app/src/main/java/com/mohsenoid/rickandmorty/data/service/network/NetworkHelperImpl.java
package com.mohsenoid.rickandmorty.data.service.network;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.List;

public class NetworkHelperImpl implements NetworkHelper {

    private String baseUrl;

    public NetworkHelperImpl(String baseUrl) {
        this.baseUrl = baseUrl;
    }

    @Override
    public String requestData(String endpoint, List<Param> params) throws IOException {
        StringBuilder response = new StringBuilder();

        URL url;
        HttpURLConnection urlConnection = null;
        try {
            StringBuilder urlBuilder = new StringBuilder();
            urlBuilder.append(baseUrl);
            urlBuilder.append(endpoint);

            if (params != null) {
                for (Param param : params) {
                    urlBuilder.append("?").append(param.getKey()).append("=").append(param.getValue());
                }
            }

            url = new URL(urlBuilder.toString());

            urlConnection = (HttpURLConnection) url.openConnection();

            InputStream inputStream = urlConnection.getInputStream();

            InputStreamReader inputStreamReader = new InputStreamReader(inputStream);


            int data = inputStreamReader.read();
            while (data != -1) {
                response.append((char) data);
                data = inputStreamReader.read();
            }
        } finally {
            if (urlConnection != null) {
                urlConnection.disconnect();
            }
        }

        return response.toString();
    }

}

File: app/src/main/kotlin/com/mohsenoid/rickandmorty/data/network/NetworkHelperImpl.kt
package com.mohsenoid.rickandmorty.data.network

import java.io.*
import java.net.URL

class NetworkHelperImpl(private val baseUrl: String) : NetworkHelper {

    override fun requestData(endpoint: String, params: List<NetworkHelper.Param>?): String {
        val urlBuilder = StringBuilder()
        urlBuilder.append(baseUrl)
        urlBuilder.append(endpoint)

        if (params != null) {
            for ((key, value) in params) {
                urlBuilder.append("?").append(key).append("=").append(value)
            }
        }

        val url = URL(urlBuilder.toString())
        val byteArrayOutputStream = ByteArrayOutputStream()
        request(url, byteArrayOutputStream)
        return String(byteArrayOutputStream.toByteArray())
    }

    override fun requestImageData(imageUrl: String, imageFile: File) {
        val url = URL(imageUrl)
        val fileOutputStream = FileOutputStream(imageFile)
        request(url, fileOutputStream)
    }

    private fun request(url: URL, output: OutputStream) {
        val urlConnection = url.openConnection()
        urlConnection.connect()
        BufferedInputStream(url.openStream(), BUFFER_SIZE).use { input ->
            val data = ByteArray(BUFFER_SIZE)
            var count: Int

            while (input.read(data).also { count = it } != -1) {
                output.write(data, 0, count)
            }
        }
        output.flush()
        output.close()
    }

    companion object {
        private const val BUFFER_SIZE = 8192
    }
}
